File: go/reject_test.go
package multipath

import (
	"context"
	"io"
	"net"
	"strings"
	"sync"
	"testing"
	"time"
)

func dialTCP(addr string) func(context.Context, int) (io.ReadWriteCloser, error) {
	return func(ctx context.Context, i int) (io.ReadWriteCloser, error) {
		return net.Dial("tcp", addr)
	}
}

// TestRejectOutOfRangeLinkReportedNotRetried: a link whose HELLO index the origin refuses is reported
// down with the origin's reason (not a bare "connection closed") and is not retried, while an
// in-range link on the same stream stays up.
func TestRejectOutOfRangeLinkReportedNotRetried(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()
	srv := Listen(ln, fastOpts(1, nil)) // origin accepts index 0 only
	go func() { _, _ = srv.Accept() }()

	var mu sync.Mutex
	var events []LinkState
	copt := fastOpts(2, dialTCP(ln.Addr().String())) // client offers index 0 and 1; 1 is out of range
	copt.OnLinkState = func(e LinkState) {
		mu.Lock()
		events = append(events, e)
		mu.Unlock()
	}
	cli, err := DialRedundant(context.Background(), copt)
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	defer cli.Close()

	waitFor(t, func() bool {
		mu.Lock()
		defer mu.Unlock()
		for _, e := range events {
			if e.Index == 1 && !e.Up && strings.Contains(e.Reason, "out of range") {
				return true
			}
		}
		return false
	}, "link 1 reported down with the origin's out-of-range reason")

	// Link 0 stays up; link 1 stays rejected and is not retried (no reconnect churn).
	waitFor(t, func() bool {
		for _, s := range cli.Stats() {
			if s.Index == 0 && s.State == "up" {
				return true
			}
		}
		return false
	}, "link 0 up")
	time.Sleep(200 * time.Millisecond)
	if r := cli.Stats()[1].Reconnects; r != 0 {
		t.Fatalf("rejected link 1 was retried %d times; it must not be", r)
	}
}

// TestRejectAllLinksFailsFast reproduces the expensive bug: the origin refuses every link, Dial still
// returns success (it only wrote HELLO), and without the fix every request hangs forever with no log.
// With it, the stream closes with the origin's reason so the next read/write fails fast.
func TestRejectAllLinksFailsFast(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()
	_ = Listen(ln, RedundantOptions{N: 0}) // origin accepts no index at all → rejects every link

	cli, err := DialRedundant(context.Background(), fastOpts(1, dialTCP(ln.Addr().String())))
	if err != nil {
		t.Fatalf("dial should still succeed (HELLO written): %v", err)
	}
	defer cli.Close()

	done := make(chan error, 1)
	go func() {
		_, e := cli.Read(make([]byte, 16))
		done <- e
	}()
	select {
	case e := <-done:
		if e == nil || !strings.Contains(e.Error(), "rejected") {
			t.Fatalf("want a reject error, got %v", e)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("Read hung instead of failing fast when every link was rejected")
	}
}

File: jvm/src/test/kotlin/app/microteams/multipath/redundant/RejectTest.kt
package app.microteams.multipath.redundant

import java.io.IOException
import java.net.ServerSocket
import java.net.Socket
import java.time.Duration
import java.util.Collections
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTimeoutPreemptively
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class RejectTest {

    private fun waitUntil(msg: String, cond: () -> Boolean) {
        val deadline = System.currentTimeMillis() + 5000
        while (System.currentTimeMillis() < deadline) {
            if (cond()) return
            Thread.sleep(5)
        }
        throw AssertionError("timed out waiting for: $msg")
    }

    private fun dialTo(port: Int): (Int) -> LinkConn? = {
        val s = Socket("127.0.0.1", port)
        s.tcpNoDelay = true
        LinkConn(s.getInputStream(), s.getOutputStream(), s)
    }

    // A link whose HELLO index the origin refuses is reported down with the origin's reason and
    // never
    // retried, while an in-range link on the same stream stays up.
    @Test
    fun outOfRangeLinkReportedNotRetried() {
        val events = Collections.synchronizedList(mutableListOf<LinkState>())
        val srvSock = ServerSocket(0)
        val server = RedundantServer(srvSock, RedundantOptions(n = 1)) // accepts index 0 only
        val client =
            RedundantStream.dial(
                RedundantOptions(
                    n = 2, // offers index 0 and 1; 1 is out of range
                    pingIntervalMs = 20,
                    deadAfterMs = 500,
                    reconnectDelayMs = 10,
                    maxDelayMs = 50,
                    onLinkState = { events.add(it) },
                ),
                dialTo(srvSock.localPort),
            )
        val accepted = server.accept()

        waitUntil("link 1 reported down with the out-of-range reason") {
            events.any { it.index == 1 && !it.up && it.reason.contains("out of range") }
        }
        waitUntil("link 0 up") { client.stats()[0].state == "up" }
        Thread.sleep(200)
        assertEquals(0, client.stats()[1].reconnects, "rejected link 1 must not be retried")

        client.close()
        accepted.close()
        server.close()
    }

    // The origin refuses every link: dial still returns (HELLO was written), but the stream then
    // closes with the reason so read fails fast instead of hanging forever.
    @Test
    fun allLinksRejectedFailFast() {
        val srvSock = ServerSocket(0)
        val server =
            RedundantServer(srvSock, RedundantOptions(n = 0)) // accepts no index → rejects all
        val client =
            RedundantStream.dial(
                RedundantOptions(n = 1, pingIntervalMs = 20, reconnectDelayMs = 10),
                dialTo(srvSock.localPort),
            )

        assertTimeoutPreemptively(Duration.ofSeconds(5)) {
            val ex = assertThrows(IOException::class.java) { client.read(ByteArray(16), 0, 16) }
            assertTrue(
                ex.message!!.contains("rejected"),
                "read should fail with the reject reason, got: ${ex.message}",
            )
        }

        client.close()
        server.close()
    }
}
